cvmfs/file_watcher_kqueue.cc

```diff
@@ -0,0 +1,164 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#include <fcntl.h>
+#include <sys/event.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <string>
+#include <vector>
+
+#include "backoff.h"
+#include "file_watcher_kqueue.h"
+#include "logging.h"
+#include "util/posix.h"
+
+namespace file_watcher {
+
+FileWatcherKqueue::FileWatcherKqueue() {}
+
+FileWatcherKqueue::~FileWatcherKqueue() {}
+
+bool FileWatcherKqueue::RunEventLoop(const FileWatcher::HandlerMap& handlers,
+                                     int control_pipe) {
+  kq_ = kqueue();
+  if (kq_ == -1) {
+    LogCvmfs(kLogCvmfs, kLogDebug, "FileWatcherKqueue - Cannot create kqueue.");
+    return false;
+  }
+
+  // Control pipe sending the stop event.
+  struct kevent watch_event;
+  EV_SET(&watch_event, control_pipe, EVFILT_READ, EV_ADD | EV_ENABLE | EV_CLEAR,
+         0, 0, 0);
+  if (kevent(kq_, &watch_event, 1, NULL, 0, NULL) == -1) {
+    LogCvmfs(kLogCvmfs, kLogDebug,
+             "FileWatcherKqueue - Could not register event with kqueue.");
+    return false;
+  }
+
+  for (FileWatcher::HandlerMap::const_iterator it = handlers.begin();
+       it != handlers.end(); ++it) {
+    RegisterFilter(it->first, it->second);
+  }
+
+  bool stop = false;
+  while (!stop) {
+    std::vector<struct kevent> triggered_events(handlers.size() + 1);
+    int nev = kevent(kq_, NULL, 0, &triggered_events[0],
+                     triggered_events.size(), NULL);
+    if (nev == -1) {
+      LogCvmfs(kLogCvmfs, kLogDebug,
+               "FileWatcherKqueue - Could not poll events.");
+      return false;
+    }
+    if (nev == 0) {
+      continue;
+    }
+    for (int i = 0; i < nev; ++i) {
+      struct kevent& current_ev = triggered_events[i];
+      if (current_ev.ident == static_cast<uint64_t>(control_pipe)) {
+        char buffer[1];
+        read(control_pipe, &buffer, 1);
+        LogCvmfs(kLogCvmfs, kLogDebug, "FileWatcherKqueue - Stopping.\n");
+        stop = true;
+        continue;
+      } else {
```

Comment:
The `else` branch could be dropped, making the control flow a bit easier to follow.